### ISSUE TITLE
Consistent id_tap grad behaviour

### ISSUE THREAD

**Issue by sbodenstein** [google/jax#5883]
Often when using `host_callback.id_tap` for logging, you do not want the gradients. The cleanest way is to just not use the result of `id_tap`. But this is unsafe (it might be optimized away), so for safety you, always need to use the result. Conceptually, the same thing can be achieved with setting the `result` argument of `id_tap`. But instead of just avoiding the Python callback for the gradient, it invokes a callback with the (useless) 0 grad:
```
import jax
from jax.experimental import host_callback

def f(x):
  x = host_callback.id_print(x, result=x, what='log')
  return x

jax.grad(f)(2.)

# what: log
# 2.
# transforms: ['jvp', 'transpose'] what: log
# 0.
# DeviceArray(1., dtype=float32)
```
Would it make sense/be possible to avoid this unnecessary callback?

**Comment by rajasekharporeddy** [comment]
Hi @sbodenstein 

It looks like this issue has been resolved in the later versions of JAX. I executed the mentioned code in Colab using JAX version 0.4.23. It avoids the Python callback for the gradient in the output.

```python
import jax
from jax.experimental import host_callback


print(jax.__version__)

def f(x):
  x = host_callback.id_print(x, result=x, what='log')
  return x

jax.grad(f)(2.)
```

Output:

```
0.4.23
what: log
2.
Array(1., dtype=float32, weak_type=True)
```

Kindly find the [gist](https://colab.research.google.com/gist/rajasekharporeddy/48dac9dda8b79342dad38b7e9963a5ca/-5883.ipynb) for reference.

Thank you

**Comment by jakevdp** [comment]
`id_print` is deprecated; going forward you should use [`jax.debug.print`](https://jax.readthedocs.io/en/latest/_autosummary/jax.debug.print.html#jax.debug.print) which does not log any gradient:
```python
import jax

def f(x):
  jax.debug.print("debug.print(x): {}", x)
  return x

jax.grad(f)(2.)
```
```
x value is 2.0
```
